### PR TITLE
feat: complete lc112, lc113, lc129 and lc437

### PR DESCRIPTION
**Pull request by janmarkusmilan**
1. Copied and pasted DFS problems into my folder
2. Completed lc112, lc113, lc129 and lc437
3. Updated README in my folder to reflect changes

### CODE REVIEW

File: janMarkusMilan/assignments/treesDFS/lc112/lc112.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    count++;
+    return recursive(node, count);
+  }
+
+What to take note of:
+  1. Since the question is checking for a sum, then we need to pass that as an argument in our recursive function (including the node)
+  2. For each node we reach, we add the value contained inside to our sum
+  3. If our sum equals the target sum at any point, then just return true
+  4. If we reach the end of a branch, we check if our sum equals the target sum
+*/
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc112/lc112.js:23]: Good approach.

File: janMarkusMilan/assignments/treesDFS/lc112/lc112.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+root.right = new TreeNode(1);
+root.left.left = new TreeNode(9);
+root.right.left = new TreeNode(10);
+root.right.right = new TreeNode(5);
+console.log(`Tree has path: ${has_path(root, 23)}`);
+console.log(`Tree has path: ${has_path(root, 16)}`);
+console.log(`Tree has path: ${has_path(root, 12)}`);
+console.log(`Tree has path: ${has_path(root, 0)}`);
+console.log(`Tree has path: ${has_path(root, null)}`);
+console.log(`Tree has path: ${has_path(root, 28)}`);
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc112/lc112.js:71]: Nice job adding custom test cases!

File: janMarkusMilan/assignments/treesDFS/lc112/lc112.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    if (DFS(node.left, currentSum) || DFS(node.right, currentSum)) {
+      return true;
+    }
+    return false;
+  }
+
+  return DFS(root, 0);
+};
+
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc112/lc112.js:51]: Don't forget to write it twice to build up your muscle memory.

File: janMarkusMilan/assignments/treesDFS/lc113/lc113.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+Things to take note:
+  1. The problem is asking for a sum, so that should be a parameter in our recursive function (as well as the node)
+  2. The output is an array of arrays, so an array should also be a parameter in our recursive function
+  3. For each node we reach, we add the value of the node to our sum
+  4. When we reach the end of a branch, we check if our sum is equal to the target sum 
+  5. We only want to push our array to our final output if we reached the end of the branch (no left or right node) and the sum equals the target
+  6. Outside that, we just recursively call the function for left and right nodes
+*/
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc113/lc113.js:35]: Good approach!

File: janMarkusMilan/assignments/treesDFS/lc113/lc113.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+sum = null;
+res = find_paths(root, sum);
+console.log(`Tree paths with sum = ${sum}:`);
+console.log(res);
+
+sum = 18;
+res = find_paths(root, sum);
+console.log(`Tree paths with sum = ${sum}:`);
+console.log(res);
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc113/lc113.js:103]: Nice job adding custom test cases!

File: janMarkusMilan/assignments/treesDFS/lc113/lc113.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    if (node.right) {
+      DFS(node.right, currentSum, path.slice());
+    }
+  }
+
+  DFS(root, 0, []);
+  return allPaths;
+};
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc113/lc113.js:74]: Don't forget to write it twice to build up your muscle memory!

File: janMarkusMilan/assignments/treesDFS/lc129/lc129.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  return recursive(node, count);
+}
+
+Things to take note:
+  1. For each value, covert them to strings and them add them together (1 and 4 -> '1' and '4' -> '1' + '4' = '14')
+  2. Another option is to multiply by 10 and then add the current value in the node (1 and 7 -> 1 * 10 = 10 + 7 = 17)
+  3. At the end, I can convert the string back to an integer and then add it
+  4. Pass sum as argument in my recursive function -> recursive (node, sum)
+  5. In my function, I would perform some operation on it
+*/
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc129/lc129.js:32]: Nice approach!

File: janMarkusMilan/assignments/treesDFS/lc129/lc129.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+root2.left = new TreeNode(3);
+root2.right = new TreeNode(1);
+root2.left.left = new TreeNode(5);
+root2.right.left = new TreeNode(2);
+root2.right.right = new TreeNode(9);
+console.log(`Total Sum of Path Numbers: ${find_sum_of_path_numbers(root2)}`);
+
+var root3 = new TreeNode(0);
+root3.left = new TreeNode(1);
+console.log(`Total Sum of Path Numbers: ${find_sum_of_path_numbers(root3)}`);
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc129/lc129.js:87]: :100:

File: janMarkusMilan/assignments/treesDFS/lc129/lc129.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+/*
+        (1)           101 + 116 + 115 = 217 + 115 = 332
+        / \
+      (0) (1)
+      /   / \
+    (1) (6) (5)
+*/
+
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc129/lc129.js:68]: You should also add a test case to check your edge cases, i.e. root === null

File: janMarkusMilan/assignments/treesDFS/lc129/lc129.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    if (!node.left && !node.right) {
+      return parseInt(sum);
+    }
+
+    return DFS(node.left, sum) + DFS(node.right, sum);
+  }
+
+  return DFS(root, 0);
+};
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc129/lc129.js:59]: Don't forget to write it twice to build up your muscle memory.

File: janMarkusMilan/assignments/treesDFS/lc437/lc437.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+How to implement DFS recursively for this question:
+  1. In our BFS loop, we should call the DFS function using the current node
+  2. Our DFS function should also have a sum parameter in order to get the sum of the current path
+  3. For each node we iterate to, we add the value of the node to our sum
+  4. What we ultimately want to return from our DFS is an integer (the number of valid paths)
+  5. We can initialize some variable, and if our sum is equal to the target sum, we increment it by 1
+  6. Then we can just recursively call the DFS function for left and right nodes
+*/
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc437/lc437.js:67]: Very detailed approach and explanation.

File: janMarkusMilan/assignments/treesDFS/lc437/lc437.js [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+root.left = new TreeNode(7);
+root.right = new TreeNode(1);
+root.left.left = new TreeNode(4);
+root.right.left = new TreeNode(10);
+root.right.right = new TreeNode(5);
+console.log(`Tree has paths: ${count_paths(root, 11)}`);
+console.log(`Tree has paths: ${count_paths(root, null)}`);
+console.log(`Tree has paths: ${count_paths(root, 19)}`);
+console.log(`Tree has paths: ${count_paths(root, 5)}`);
+console.log(`Tree has paths: ${count_paths(root, 100)}`);
```
Comment (knrodriguez) [inline janMarkusMilan/assignments/treesDFS/lc437/lc437.js:137]: :100:

**Review by knrodriguez — APPROVED** [review]

**Comment by knrodriguez** [comment]
Nice work! Your approaches, while not step-by-step, are still clear. You added custom test cases, but make sure you add at least that tests your edge cases. Also, writing it twice will help you build up muscle memory should you encounter the problem again in the future.
